proposals/p3720.md

```diff
@@ -0,0 +1,1258 @@
+
+With these two ingredients, `b.Search(32)` is equivalent to
+`b.(String.Search)(32)`, which is then equivalent to
+`b.ptr->(String.Search)(32)` or `b.ptr->Search(32)` (since `b.ptr` has type
+`String*`).
+
+### Other uses of `extend api`
+
+The [`extend api` mechanism](#extend-api), allowing lookup to find names from
+another type, has other uses beyond [member forwarding](#member-forwarding).
+
+A class could extend the API of a class it implicitly converts to, see
+[inheritance and other implicit conversions](#inheritance-and-other-implicit-conversions).
+For example, imagine we have a class representing an integer in a restricted
+range that can implicitly convert to an integer value.
+
+```carbon
+class InRange(Low:! i32, High:! i32) {
+  var value: i32;
+  impl as ImplicitAs(i32) {
+    fn Convert[self: Self]() -> i32 { return self.value; }
+  }
+  extend api i32;
+}
+```
+
+By including `extend api i32`, `InRange` gains support for any non-`addr`
+methods on `i32`, like perhaps `Abs`.
+
+Or a class could have members specifically intended for use by another class, as
+[extension methods](https://github.com/carbon-language/carbon-lang/issues/1345)
+-- effectively acting as mixin except it can't add member variables.
+
+The examples so far use `extend api` between two classes, but we also allow it
+with interfaces and named constraints.
+
+For example, a class can `extend api` of an interface it
+[conditionally conforms to](/docs/design/generics/details.md#conditional-conformance),
+as in:
+
+```carbon
+interface A {
+  fn F();
+}
+class C(T:! type) {
+  extend api A;
+}
+impl C(i32) as A { fn F() { ... } }
+```
+
+Here, the class `C(T)` implements `A` for some values of `T`, such as `i32`.
+Rather than manually individually aliasing the members of `A` in `C`, to make
+them available as part of the API of `C`, so that `C(i32).F()` is valid, an
+`extend api` declaration includes all of them at once.
+
+Another use case is that an interface can `extend api` of another interface. In
+this example,
+
+```carbon
+interface A {
+  fn F();
+}
+interface B {
+  extend api A;
+  fn G();
+}
+```
+
+`B.F` would be an alias for `A.F`, but without any implied
+`require Self impls A`, in contrast with a plain `extend A`. So `extend I;` is
+equivalent to `require Self impls I; extend api I;`.
+
+Lastly, an interface could `extend api` a class. This could be done to add
+something that acts like `final` functions to the interface, using
+[extension methods](https://github.com/carbon-language/carbon-lang/issues/1345),
+as in:
+
+```carbon
+class Helpers {
+  fn F[T:! type, self: T]() { DoStuffWith(self); }
+  // ...
+}
+interface Iface {
+  extend api Helpers;
+  fn G();
+  // ...
+}
+class C {
+  extend impl as Iface;
+}
+fn Test(c: C) {
+  // Calls Helpers.F from extended class Helpers
+  c.F();
+}
+```
+
+Unlike `final` functions in an interface, this approach defines all the helpers
+in a separate entity that could be used to extend more than one type.
```

Comment:
Looking at this again, you are correct. This can indeed be achieved by extending an interface with a set of final methods. I have made this change in #3802.